nixos/modules/services/monitoring/staytus.nix

```diff
@@ -0,0 +1,370 @@
+          ${concatStrings (mapAttrsToList (name: value: "--set ${name} '${value}' ") staytusEnv)} \
+          --set PATH '${lib.makeBinPath (with pkgs; [ nodejs ])}:$PATH' \
+          --set RAKEOPT '-f /run/staytus/Rakefile' \
+          --run 'env' \
+          --run 'cd /run/staytus'
+      makeWrapper $out/bin/staytus-bundle $out/bin/staytus-rake \
+          --add-flags "exec rake"
+      makeWrapper $out/bin/staytus-rake $out/bin/staytus-jobs \
+          --add-flags "jobs:work"
+      makeWrapper $out/bin/staytus-bundle $out/bin/staytus-puma \
+          --add-flags "exec puma" \
+          --add-flags "--environment production" \
+          --add-flags "-C /run/staytus/config/puma.rb" \
+          --add-flags "--redirect-stdout ${cfg.stateDir}/log/puma.log" \
+          --add-flags "--redirect-stderr ${cfg.stateDir}/log/puma.error.log" \
+          --add-flags "--redirect-append" \
+          --add-flags "--pidfile /run/staytus/pids/puma.pid"
+     '';
+  };
+
+in {
+
+  options = {
+    services.staytus = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Enable the staytus service.
+        '';
+      };
+
+      stateDir = mkOption {
+        type = types.str;
+        default = "/var/staytus";
+        description = "The state directory, logs and plugins are stored here";
+      };
+
+      extraConfig = mkOption {
+        type = types.lines;
+        default = "";
+        description = "Extra configuration in configuration.yml";
+      };
+
+      themes = mkOption {
+        type = types.attrsOf types.path;
+        default = {};
+        description = "Set of themes";
+      };
+
+      theme = mkOption {
+        type = types.str;
+        default = "default";
+        description = "The theme to activate";
+      };
+
+      #databaseType = mkOption {
+      #  type = types.str;
+      #  default = "postgresql";
+      #  description = "Type of database";
+      #};
+
+      databaseHost = mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        description = "Database hostname";
+      };
+
+      databasePassword = mkOption {
+        type = types.str;
+        default = "";
+        description = "Database user password";
+      };
+
+      databaseName = mkOption {
+        type = types.str;
+        default = "staytus";
+        description = "Database name";
+      };
+
+      databaseUsername = mkOption {
+        type = types.str;
+        default = "staytus";
+        description = "Database user";
+      };
+
+      smtpHost = mkOption {
+        type = types.str;
+        default = "smtp.deliverhq.com";
+        description = "SMTP hostname";
+      };
+
+      smtpPort = mkOption {
+        type = types.str;
+        default = "25";
+        description = "SMTP port";
+      };
+
+      smtpPassword = mkOption {
+        type = types.str;
+        default = "";
+        description = "SMTP user password";
+      };
```

Comment:
When `smtpPassword` and `secretKeyBase` will end world-readable up in the nix store. This might be not desired depending on the deployment. Until https://github.com/NixOS/nix/issues/8 is solved we add the option to store secrets in a dedicated file, that is copied/loaded at runtime by the service. As the configuration is created at build time, substitution with sed can be used to replace a password place holder with a concrete value.

Reply:
I'll look into it, thanks.

Reply:
How does one handle the password file with NixOps?

Reply:
I think it might be possible to use unix socket authentication to reset the password to something random every time the service is started.

So the staytus user would connect to mysql via unix socket (along the lines of this: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/databases/mysql.nix#L368), then add an identification using a password that is auto generated, which is then set in an env var, and picked up by rails.

On the face of it, it seems possible, so I'll have a closer look next week.